--- a/init/control.c
+++ b/init/control.c
@@ -1200,12 +1200,16 @@
 
 		/* this can be a null value */
 		instance = job_details[1];
-	} else if (getpid () == 1) {
-		nih_dbus_error_raise_printf (
-			DBUS_INTERFACE_UPSTART ".Error.PermissionDenied",
-			_("Not permissible to modify PID 1 job environment"));
-		return -1;
-	}
+	}
+/*
+ * This block prevents the set-env --global from working
+ *	  else if (getpid () == 1) {
+ *		nih_dbus_error_raise_printf (
+ *			DBUS_INTERFACE_UPSTART ".Error.PermissionDenied",
+ *			_("Not permissible to modify PID 1 job environment"));
+ *		return -1;
+ *	} 
+ */
 
 	if (! control_check_permission (message)) {
 		nih_dbus_error_raise_printf (
@@ -1305,12 +1309,16 @@
 
 		/* this can be a null value */
 		instance = job_details[1];
-	} else if (getpid () == 1) {
-		nih_dbus_error_raise_printf (
-			DBUS_INTERFACE_UPSTART ".Error.PermissionDenied",
-			_("Not permissible to modify PID 1 job environment"));
-		return -1;
-	}
+	}
+/*
+ *	This block prevents initctl set-env --global from working
+ *	  else if (getpid () == 1) {
+ *		nih_dbus_error_raise_printf (
+ *			DBUS_INTERFACE_UPSTART ".Error.PermissionDenied",
+ *			_("Not permissible to modify PID 1 job environment"));
+ *		return -1;
+ *	} 
+ */
 
 	/* Verify that job name is valid */
 	if (job_name && ! strlen (job_name)) {
@@ -1559,12 +1567,16 @@
 
 		/* this can be a null value */
 		instance = job_details[1];
-	} else if (getpid () == 1) {
-		nih_dbus_error_raise_printf (
-			DBUS_INTERFACE_UPSTART ".Error.PermissionDenied",
-			_("Not permissible to modify PID 1 job environment"));
-		return -1;
-	}
+	}
+/*
+ *	This block prevents initctl set-env --global from working
+ *	 else if (getpid () == 1) {
+ *		nih_dbus_error_raise_printf (
+ *			DBUS_INTERFACE_UPSTART ".Error.PermissionDenied",
+ *			_("Not permissible to modify PID 1 job environment"));
+ *		return -1;
+ *	}
+ */
 
 	if (! control_check_permission (message)) {
 		nih_dbus_error_raise_printf (
