Index: git/CMakeLists.txt
===================================================================
--- git.orig/CMakeLists.txt	2012-07-05 22:40:54.000000000 -0700
+++ git/CMakeLists.txt	2012-07-05 22:41:38.000000000 -0700
@@ -117,6 +117,7 @@
 add_linker_flags("-Wl,--no-undefined")
 # XXX_ The libraries that come with gcc fail when --no-allow-shlib-undefined is specified
 #add_linker_flags("-Wl,--no-allow-shlib-undefined")
+add_linker_flags("-Wl,-lrt")
 
 
 add_subdirectory(libpowerd)
