commit 62c0bb9535663d16ee63a3945009abb290083a64
Author: Iyad Qumei <iyadkq@gmail.com>
Date:   Tue Jul 17 10:17:42 2012 -0700

    use trusted cert dirs

diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index 43a5a4f..f46174a 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -1547,6 +1547,9 @@ SSL_CTX *SSL_CTX_new(SSL_METHOD *meth)
 #endif
 #endif
 
+        // force each connection to know about the common trusted cert cache:
+        SSL_CTX_load_verify_locations(ret,NULL,"/var/ssl/trustedcerts");
+
 	return(ret);
 err:
 	SSLerr(SSL_F_SSL_CTX_NEW,ERR_R_MALLOC_FAILURE);
commit 303eeda16ae509057fb17531c526de3b711f7e10
Author: Iyad Qumei <iyadkq@gmail.com>
Date:   Tue Jul 17 10:18:49 2012 -0700

    NOV-99171

diff --git a/crypto/cryptlib.h b/crypto/cryptlib.h
index fc249c5..00d4221 100644
--- a/crypto/cryptlib.h
+++ b/crypto/cryptlib.h
@@ -81,7 +81,8 @@ extern "C" {
 
 #ifndef OPENSSL_SYS_VMS
 #define X509_CERT_AREA		OPENSSLDIR
-#define X509_CERT_DIR		OPENSSLDIR "/certs"
+//#define X509_CERT_DIR		OPENSSLDIR "/certs"
+#define X509_CERT_DIR		"/var/ssl/trustedcas"
 #define X509_CERT_FILE		OPENSSLDIR "/cert.pem"
 #define X509_PRIVATE_DIR	OPENSSLDIR "/private"
 #else

