diff --git a/home/leppaki/projects/webos/codes/github/webos/BUILD-qemux86/work/i586-palm-linux/qt4-palm-4.8.2.02-0.33/git/src/gui/painting/qpixmapatlas.cpp b/home/leppaki/tmp/qpixmapatlas.cpp
index c4ba275..672aa3d 100644
--- a/src/gui/painting/qpixmapatlas.cpp
+++ b/src/gui/painting/qpixmapatlas.cpp
@@ -33,7 +33,7 @@
 #include <QtCore/QTemporaryFile>
 
 #include <QDebug>
-
+#include <unistd.h>
 #if defined(QT_WEBOS)
 
 
