diff -urN alljoyn-3.4.2-src-org/alljoyn_core/src/Makefile alljoyn-3.4.2-src-patch/alljoyn_core/src/Makefile
--- alljoyn-3.4.2-src-org/alljoyn_core/src/Makefile	2013-11-09 10:03:15.000000000 +0900
+++ alljoyn-3.4.2-src-patch/alljoyn_core/src/Makefile	2013-11-13 13:50:41.721742739 +0900
@@ -15,6 +15,8 @@
 
 STATIC_LIB = liballjoyn.a
 DYNAMIC_LIB = liballjoyn.so
+MAJOR = 0
+MINOR = 1.0
 
 .PHONY: all clean
 
@@ -47,8 +49,10 @@
 corelib: $(ALLJOYN_CC_OBJS) $(ALLJOYN_C_OBJS)
 	$(AR) r $(STATIC_LIB) $(ALLJOYN_CC_OBJS) $(ALLJOYN_C_OBJS) $(COMMON_OBJS)
 	cp $(STATIC_LIB) $(INSTALLDIR)/dist/lib
-	$(CC) -o $(DYNAMIC_LIB) $(CXXFLAGS) $(CPPDEFINES) $(LINKFLAGS) $(INCLUDE)  $(ALLJOYN_CC_OBJS) $(ALLJOYN_C_OBJS) $(COMMON_OBJS) $(LIBS)
-	cp $(DYNAMIC_LIB) $(INSTALLDIR)/dist/lib
+	$(CC) -Wl,-soname,$(DYNAMIC_LIB).$(MAJOR) -o $(DYNAMIC_LIB).$(MAJOR).$(MINOR) $(CXXFLAGS) $(CPPDEFINES) $(LINKFLAGS) $(INCLUDE)  $(ALLJOYN_CC_OBJS) $(ALLJOYN_C_OBJS) $(COMMON_OBJS) $(LIBS)
+	cp $(DYNAMIC_LIB).$(MAJOR).$(MINOR) $(INSTALLDIR)/dist/lib
+	cd $(INSTALLDIR)/dist/lib && ln -sf $(DYNAMIC_LIB).$(MAJOR).$(MINOR) $(DYNAMIC_LIB)
+	cd $(INSTALLDIR)/dist/lib && ln -sf $(DYNAMIC_LIB).$(MAJOR).$(MINOR) $(DYNAMIC_LIB).$(MAJOR)
 
 clean:
 	@rm -f *.o *~ $(STATIC_LIB) $(DYNAMIC_LIB) $(OS_SRC_DIR)/*.o $(OS_SRC_DIR)/linux/*.o
diff -urN alljoyn-3.4.2-src-org/alljoyn_js/Makefile alljoyn-3.4.2-src-patch/alljoyn_js/Makefile
--- alljoyn-3.4.2-src-org/alljoyn_js/Makefile	2013-11-09 10:03:47.000000000 +0900
+++ alljoyn-3.4.2-src-patch/alljoyn_js/Makefile	2013-11-13 13:50:52.745490137 +0900
@@ -20,9 +20,9 @@
 
 all: distclean dist
 
-CC=gcc
-CXX=g++
-AR=ar
+#CC=gcc
+#CXX=g++
+#AR=ar
 
 # build configuration
 ifeq "$(CPU)" "x86"
@@ -77,9 +77,9 @@
 
 ifeq "$(CPU)" "x86"
 # Force 32-bit builds
-    CXXFLAGS+=-m32 -pthread
-    CFLAGS+=-m32 -pthread
-    LINKFLAGS+=-m32 -pthread
+    CXXFLAGS+=-fPIC -pthread
+    CFLAGS+=-fPIC -pthread
+    LINKFLAGS+=-fPIC -pthread
 else
     ifeq "$(CPU)" "x86_64"
         CXXFLAGS+=-m64 -fPIC -pthread
@@ -109,7 +109,7 @@
     LIBS += -lrt -lstdc++ -lpthread -lcrypto -lssl
 endif
 
-INSTALLDIR = $(PWD)/../build/$(OS)/$(CPU)/$(VARIANT)
+INSTALLDIR = $(PWD)/../build/
 ALLJOYNLIB = $(INSTALLDIR)/dist/lib/liballjoyn.a
 AJDAEMONLIB = $(INSTALLDIR)/dist/lib/libajdaemon.a
 BUNDLED_OBJ = $(INSTALLDIR)/dist/lib/BundledDaemon.o
diff -urN alljoyn-3.4.2-src-org/common/Makefile alljoyn-3.4.2-src-patch/common/Makefile
--- alljoyn-3.4.2-src-org/common/Makefile	2013-11-09 10:04:55.000000000 +0900
+++ alljoyn-3.4.2-src-patch/common/Makefile	2013-11-13 13:51:00.300767159 +0900
@@ -23,17 +23,10 @@
 	cd crypto/$(CRYPTO); make;
 
 	@mkdir -p $(INSTALLDIR)/dist/inc/qcc
-	cp inc/qcc/Log.h \
-       inc/qcc/ManagedObj.h \
-       inc/qcc/String.h \
-       inc/qcc/atomic.h \
-       inc/qcc/SocketWrapper.h \
-       inc/qcc/platform.h $(INSTALLDIR)/dist/inc/qcc
+	cp inc/qcc/*.h $(INSTALLDIR)/dist/inc/qcc
 
 	@mkdir -p $(INSTALLDIR)/dist/inc/qcc/$(OS_GROUP)
-	cp inc/qcc/${OS_GROUP}/atomic.h \
-       inc/qcc/${OS_GROUP}/platform_types.h \
-       inc/qcc/${OS_GROUP}/unicode.h $(INSTALLDIR)/dist/inc/qcc/$(OS_GROUP)
+	cp inc/qcc/${OS_GROUP}/* $(INSTALLDIR)/dist/inc/qcc/$(OS_GROUP)
 
 clean:
 	cd src; make clean;
diff -urN alljoyn-3.4.2-src-org/Makefile alljoyn-3.4.2-src-patch/Makefile
--- alljoyn-3.4.2-src-org/Makefile	2013-11-09 10:03:15.000000000 +0900
+++ alljoyn-3.4.2-src-patch/Makefile	2013-11-13 13:50:29.006979374 +0900
@@ -18,11 +18,11 @@
 
 .PHONY: all dist clean distclean
 
-all: distclean dist
+all: dist
 
-CC=gcc
-CXX=g++
-AR=ar
+#CC=gcc
+#CXX=g++
+#AR=ar
 
 # build configuration
 ifeq "$(CPU)" "x86"
@@ -51,12 +51,15 @@
 endif
 endif
 
-ICE?=on
-ifeq "$(OS)" "linux"
-BT?=on
-else
+#ICE?=on
+#ifeq "$(OS)" "linux"
+#BT?=on
+#else
+#BT?=off
+#endif
+
 BT?=off
-endif
+ICE?=off
 
 ifeq "$(OS)" "linux"
     # Linux specific flags
@@ -83,8 +86,8 @@
 
 ifeq "$(CPU)" "x86"
 # Force 32-bit builds
-    CXXFLAGS+=-m32
-    CFLAGS+=-m32
+    CXXFLAGS+=-fPIC
+    CFLAGS+=-fPIC
 else
     ifeq "$(CPU)" "x86_64"
         CXXFLAGS+=-m64 -fPIC
@@ -110,12 +113,12 @@
 
 # Platform specifics system libs
 ifeq "$(OS)" "linux"
-    LIBS += -lrt -lstdc++ -lpthread -lcrypto -lssl
+    LIBS += -lrt -lstdc++ -lpthread -lcrypto -lssl -ldl -lm
 endif
 
 CRYPTO = openssl
 
-INSTALLDIR = $(PWD)/build/$(OS)/$(CPU)/$(VARIANT)
+INSTALLDIR = $(PWD)/build/
 ALLJOYNLIB = $(INSTALLDIR)/dist/lib/liballjoyn.a
 COMMONDIR = $(PWD)/common
 
@@ -132,15 +135,15 @@
 	@mkdir -p $(INSTALLDIR)/dist/samples
 	cd common; make;
 	cd alljoyn_core; make;
+	cd alljoyn_js; make;
 
 clean:
 	@rm -f $(JUNK)
 	cd common; make clean;
 	cd alljoyn_core; make clean;
+	cd alljoyn_js; make clean;
 
 distclean: clean
-	@rm -rf $(INSTALLDIR)/dist/lib
-	@rm -rf $(INSTALLDIR)/dist/bin
-	@rm -rf $(INSTALLDIR)/dist/inc
-	@rm -rf $(INSTALLDIR)/dist/samples
+	@rm -rf $(INSTALLDIR)/dist/*
+	@rm -rf $(INSTALLDIR)/dist/
 
