diff --git a/Source/WebCore/platform/webos/LunaServiceMgr.cpp b/Source/WebCore/platform/webos/LunaServiceMgr.cpp
index 0452a65..11b8267 100644
--- a/Source/WebCore/platform/webos/LunaServiceMgr.cpp
+++ b/Source/WebCore/platform/webos/LunaServiceMgr.cpp
@@ -2,7 +2,7 @@
 #include "config.h"
 #if ENABLE(PALM_SERVICE_BRIDGE)
 #include "LunaServiceMgr.h"
-
+#include <unistd.h>
 #include "PlatformString.h"
 #include <glib.h>
 #include <lunaservice.h>
